bvm/ballerina-core/src/main/java/org/ballerinalang/bre/bvm/CPU.java

```diff
@@ -4127,6 +4168,110 @@ private static boolean checkIsType(BValue sourceVal, BType targetType) {
         return checkIsType(sourceType, targetType, new ArrayList<>());
     }
 
+    private static boolean checkIsLikeType(BValue sourceValue, BType targetType) {
+        if (checkIsType(sourceValue, targetType)) {
+            return true;
+        }
+
+        switch (targetType.getTag()) {
+            case TypeTags.RECORD_TYPE_TAG:
+                return checkIsLikeRecordType(sourceValue, targetType);
+            case TypeTags.JSON_TAG:
+                return checkIsLikeJSONType(sourceValue, targetType);
+            case TypeTags.MAP_TAG:
+                return checkIsLikeMapType(sourceValue, targetType);
+            case TypeTags.ARRAY_TAG:
+                return checkIsLikeArrayType(sourceValue, targetType);
+            case TypeTags.TUPLE_TAG:
+                return checkIsLikeTupleType(sourceValue, targetType);
+            case TypeTags.ANYDATA_TAG:
+                return isAssignable(sourceValue.getType(), targetType, new ArrayList<>());
+            default:
+                return false;
+        }
+    }
+
+    private static boolean checkIsLikeTupleType(BValue sourceValue, BType targetType) {
+        BRefType<?>[] arrayValues = ((BRefValueArray) sourceValue).getValues();
+        for (int i = 0; i < ((BRefValueArray) sourceValue).size(); i++) {
+            if (!checkIsLikeType(arrayValues[i], ((BTupleType) targetType).getTupleTypes().get(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean checkIsLikeArrayType(BValue sourceValue, BType targetType) {
+        BType arrayElementType = ((BArrayType) targetType).getElementType();
+        BRefType<?>[] arrayValues = ((BRefValueArray) sourceValue).getValues();
+        for (int i = 0; i < ((BRefValueArray) sourceValue).size(); i++) {
+            if (!checkIsLikeType(arrayValues[i], arrayElementType)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean checkIsLikeMapType(BValue sourceValue, BType targetType) {
+        for (Object mapEntry : ((BMap) sourceValue).values()) {
+            if (!checkIsLikeType((BValue) mapEntry, ((BMapType) targetType).getConstrainedType())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean checkIsLikeJSONType(BValue sourceValue, BType targetType) {
+        if (((BJSONType) targetType).getConstrainedType() != null) {
+            return checkIsLikeType(sourceValue, ((BJSONType) targetType).getConstrainedType());
+        } else if (sourceValue.getType().getTag() == TypeTags.ARRAY_TAG) {
+            BRefType<?>[] arrayValues = ((BRefValueArray) sourceValue).getValues();
+            for (int i = 0; i < ((BRefValueArray) sourceValue).size(); i++) {
+                if (!checkIsLikeType(arrayValues[i], targetType)) {
+                    return false;
+                }
+            }
+        }
+        return true;
```

Comment:
should we check the value is a BMap value, and each value in the map is compatible with JSON?

Reply:
I don't see there is a need for this.